### PR TITLE
BUG: Check whether the datatype is of Datetime before typecasting as usual.

### LINKED ISSUE #19267: np.allclose fails with dtype datetime64 (latest release candidate)
`np.allclose` fails when comparing equal arrays with dtype `

### Reproducing code example:
```python
import numpy as np
a = np.zeros(1, dtype='datetime64[ns]')
np.allclose(a, a)
```
fails with:

### Error message:

```
*** TypeError: The DType <class 'numpy._FloatAbstractDType'> could not be promoted by <class 'numpy.dtype[datetime64]'>. This means that no common DType exists for the given inputs. For example they cannot be stored in a single array unless the dtype is `object`. The full list of DTypes is: (<class 'numpy.dtype[datetime64]'>, <class 'numpy._FloatAbstractDType'>)
```

### NumPy/Python version information:

1.21.0rc2 3.9.4 | packaged by conda-forge | (default, May 10 2021, 22:13:33) 
[GCC 9.3.0]

### PR DESCRIPTION
**Pull request by Dinkarkumar**
Closes #19267, np.allclose fails with dtype datetime64

This is my first contribution, Seeking guidance and support from such huge  community.

### CODE REVIEW

**Comment by Dinkarkumar** [comment]
Reproducing code example:
Earlier the following code snippet used to give TypeError:-
import numpy as np
a = np.zeros(1, dtype='datetime64[ns]')
np.allclose(a, a)

Now after code changes the typecasting would not be possible and the np.allclose(a,a) would run for even dtype = 'datetime64[ns]'
@mattip , i am not yet convinced of why do you ask me to revert the chnanges. Let's discuss and resolve the issue.

**Comment by mattip** [comment]
I asked only that you revert the changes to `README.md`. You still did not add a test, the place to add it would be `numpy/core/tests/test_numeric.py`.

File: README.md [hunk 1]
```diff
@@ -60,7 +60,3 @@ mailing list. You are very welcome to join.
 If you are new to contributing to open source, [this
 guide](https://opensource.guide/how-to-contribute/) helps explain why, what,
 and how to successfully get involved.
-
```
Comment (rkern) [inline README.md:63]: This still needs to be reverted.

Comment (Dinkarkumar) [reply README.md:63]: @rkern I guess that portion was prior to the starting of this thread.

**Comment by Dinkarkumar** [comment]
Sorry for such immature execution of my task. I am totally new to this environment. Hope to learn and seeking guidance from you all.

**Comment by mattip** [comment]
I think you might want to back out all the changesets other than abe9ff0fdddfa9841131478 and then add the test in a second changeset. Your PR should only consist of 
- a one-line change to `numpy/core/numeric.py`
- a test added/modified in numpy/core/tests/test_numeric.py, which should also be at the most 10 lines.

**Comment by seberg** [comment]
I am going to close this.  The fix is likely right, but its a mess unfortunately and old.  It also needs a rebase anyway, creating a new PR from scratch seems just as well (which anyone is very welcome to do!).

**Pull request closed without merging**